### PR TITLE
MediaRanker Jessie

### PR DESCRIPTION
**Pull request by jessiezhang2017**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I wrote a model  method to select the item with the most votes
Describe how you approached testing that model method. What edge cases did you come up with? | I first test if the method will return an instance of Work.  Then I test a normal case, to make sure the right item will be selected.  I come up an edge case to see if two books with the same votes, how the system should do. I just set up to pick the first item in the array, 
What are `session' `flash`? What is the difference between them? | flash is a special type of hash that we utilize to send one-time messages from controllers to our views. session is a special type of hash-like-object that we utilize to keep track of data throughout a users "session", which normally ends when they close their browser. different between session and flash is data in session will not go away after the next request-response cycle.
Describe a controller filter you wrote. | I wrote a controller filter to find current user
What was one thing that you gained more clarity on through this assignment? | how to set up foreign keys
what is the Heroku URL of your deployed application? | https://mediarankerjz.herokuapp.com
Do you have any recommendations on how we could improve this project for the next cohort? |

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+Rails.application.routes.draw do
+  root 'works#home'
+
+  post 'users/login', to: 'users#login', as: 'login'
+  get 'users/login', to: 'users#new'
+  delete 'users/destroy', to: 'users#destroy', as: 'logout'
+
+  resources :works
+  resources :users, only:[:index, :show]
```
Comment (dHelmgren) [inline config/routes.rb:9]: Nice job using only the routes you need!

File: app/views/works/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+          <th>Created By</th>
+          <th width=100>Published</th>
+          <th width=100>Upvote</th>
+        </tr>
+      </thead>
+
+      <tbody>
+        <%@books.each do |book|%>
+          <tr>
+           <td><%= book.votes.length %></td>
```
Comment (dHelmgren) [inline app/views/works/index.html.erb:48]: You have the same code to show a list of works repeated 3 times. Could you use a view partial or a loop to DRY this up?

File: app/controllers/works_controller.rb [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+  before_action :find_work, only: [:show, :edit, :update, :destroy]
+
+  def home
+
+  end
+
+  def index
+    @albums = Work.album_list
+    @books = Work.book_list
+    @movies = Work.movie_list
```
Comment (dHelmgren) [inline app/controllers/works_controller.rb:11]: This could be DRYer with the use of categories!

File: app/models/vote.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+class Vote < ApplicationRecord
+  belongs_to :user
+  belongs_to :work
```
Comment (dHelmgren) [inline app/models/vote.rb:3]: You can use the `dependent: :destroy` property to allow your votes to be deleted alongside your Works. That way you don't get stuck with Works you can never delete.

File: app/models/work.rb [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+    if !top_book.nil?
+      top << top_book
+    end
+
+    if !top_movie.nil?
+      top << top_movie
+    end
+
+    spotlight = top.max_by {|e| e.votes.count}
```
Comment (dHelmgren) [inline app/models/work.rb:42]: This is nice and clean and I love it!

**Comment by dHelmgren** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene | Good!
Comprehension questions	| Good!
**General** |
Rails fundamentals (RESTful routing, use of named paths) | Yes!
Views are well-organized (DRY, use of semantic HTML, use of partials) | Could be a little DRYer, see comments
Errors are reported to the user | Yes
Business logic lives in the models | Yes!
Models are thoroughly tested, including relations, validations and any custom logic | yes!
**Wave 1 - Media** |
Splash page shows the three media categories | Yes!
Basic CRUD operations on media are present and functional | Yes, but delete could be improved, see comments
**Wave 2 - Users and Votes** |
Users can log in and log out | yes
The ID of the current user is stored in the session | yes
A user cannot vote for the same media more than once | yes
All media lists are ordered by vote count | yes
Splash page contains a media spotlight | yes
**Wave 3 - Users and Votes** |
Media pages contain lists of voting users | yes
Individual user pages and the user list are present | yes
**Optional - Styling** |
Bootstrap is used appropriately | Yes
Look and feel is similar to the original | yes
**Overall** | Great job overall! Your implementation matches the demo site very closely, and I would say the learning goals for this assignment were definitely met. There are a few places where things could be cleaned up, which I've tried to outline below, but in general I am quite happy with this submission. Keep up the hard work!
